docs/discussion/Cosign Upgrade Discussion 2024.md

```diff
@@ -0,0 +1,662 @@
+        - ~~What happens if keys + certs are left on disk due to Ratify failure?~~
+        - ~~Keys + certs will be able to be accessed by any process with filesystem access~~
+        - ~~CLI scenario will require fetching + deleting certs + keys off disk per invocation~~
+      - ~~Existing cert providers used by notation will need to be updated to read/write certs from disk~~
+      - ~~External plugins will need to be aware of namespace context in order to only access resources in namespaced directory~~
+  - ~~Option 3 (NOT recommended): Add certificates + keys to shared cache~~
+    - ~~Define two new cache keys partitioned by namespace~~
+    - ~~CANNOT use the ristretto in-memory cache as it is since external cosign process must be able to access it~~
+      - ~~This would require hard dependency on HA mode which is a bad experience. Why should a user need HA distributed cache just for Cosign?~~
+      - ~~Implement a new cache provider for cache sharing between processes on same host~~
+        - ~~NOT trivial~~
+        - ~~Require a new http cache server and clients~~
+    - ~~Cache provider initialization in cosign plugin implementation~~
+    - ~~One advantage is that not only will cert + key sharing be solved but also oras store credentials can easily shared which is currently not possible for external plugins~~
+- Extend Inline Certificate Provider
+  - add new `type` field to spec: "key" or "certificate". If not provided, default to "certificate" for backwards compatability. 
+- Extend Azure Key Vault Certificate Provider
+  - use `GetKeys` API to fetch keys from configured key vault
+  - add new `keys` field to spec
+- Multitenancy considerations
+  - Need a separate keys map for namespaced and global certificate store keys
+
+### **Option 2: Renaming CertificateStore to KeyManagementProvider** SELECTED IMPLEMENTATION
+- Same as Option 1 except we introduce a new CRD that will replace `CertificateStore` due to potential naming confusion
+- Is the existing name confusing enough to warrant changing to KeyManagementProvider?
+- How would we support CRD name change?
+  - Introduce new `KeyManagementProvider` CRD resource and then deprecate `CertificateStore`
+
+#### Proposed Config Changes
+
+Compared to the `CertificateStore`, the `KeyManagementProvider` config spec could be updated to be more flexible. A new `name`` field will be used only in CLI scenarios to mirror CRD name functionality as a unique identifier. This enables multiple KMS of same type to be used.
+
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: KeyManagementProvider
+metadata:
+  name: ratify-notation-inline-cert-kmprovider
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  type: inline
+  parameters:
+    contentType: key
+    value: |
+      ---------- BEGIN RSA KEY ------------
+      ******
+      ---------- END RSA KEY ------------
+```
+
+CLI Config
```

Comment:
Where is this CLI config JSON file located and how could users apply the changes to Ratify?

Reply:
Yes that's correct. I will add a clarification about this.